### ISSUE TITLE
Breaking out of a bam_plp_auto() loop leaks memory

### ISSUE THREAD

**Issue by jkbonfield** [samtools/htslib#299]
It knows it too...

```
    while ((plp = bam_plp_auto(iter, &tid, &pos, &n_plp))) {
         //...
        if (h_iter) {
            if (pos < h_iter->beg)
                continue;
            if (pos >= h_iter->end)
                break; //htslib bug: causes memory leak in bam_plp_auto.
        }
        putchar(base);
    }
```

The break statement there causes this warning "***[bam_plp_destroy] memory leak: 1. Continue anyway***".  Removing the break (esides giving incorrect output in my code) avoids the warning.  It also doesn't occur if the break happens after having ended all in-flight sequences, eg in a zero coverage region.

If we know it leaks, we ought to fix it! It looks a bit amateurish even admitting we leak memory to the user.

**Comment by jkbonfield** [comment]
Hmm, possibly I need to call bam_plp_reset(iter) prior to any bam_plp_destroy(iter) call, in which case samtools needs to make liberal use of this tool as I've certainly seen this warning crop up there too.  I forget where - I'll have a hunt and try and reproduce it, unless it's something we've known about and already fixed in samtools source?

**Comment by jmarshall** [comment]
It has [ever been thus](https://github.com/samtools/samtools/blob/0.1.1/bam_pileup.c#L142-L151).

The `mempool_t` in `bam_plp_destroy()` is similar to the one in _klist.h_.  It's a bit funky in that `mp_destroy()` only deallocates items in the pool that have already been `mp_free()ed`; `cnt` is a count of how many items have not been freed.

There is some hope that all we need to do is `mp_free` the whole linked list starting at `iter->head` rather than freeing just `iter->head`, but it's unclear whether `dummy` is always/sometimes/never part of the list…

**Comment by jkbonfield** [comment]
I found again the case where samtools spits out these warning messages; misuse of commands, such as feeding in unsorted bam files to mpileup:

```
@ seq3a[work/bam_fix_...]; samtools mpileup test.bam
[mpileup] 1 samples in 1 input files
<mpileup> Set max per-file depth to 8000
foo     10      N       1       ^]a     ~
foo     11      N       1       c       ~
foo     12      N       1       c       ~
...
[bam_pileup_core] the input is not sorted (chromosomes out of order)
[bam_plp_destroy] memory leak: 1. Continue anyway.
@ seq3a[work/bam_fix_...];
```

A trivial patch to bam_plp_destroy fixes this:

```
diff --git a/sam.c b/sam.c
index 7ebf29f..013e8d9 100644
--- a/sam.c
+++ b/sam.c
@@ -1620,6 +1620,7 @@ void bam_plp_init_overlaps(bam_plp_t iter)

 void bam_plp_destroy(bam_plp_t iter)
 {
+    bam_plp_reset(iter);
     if ( iter->overlaps ) kh_destroy(olap_hash, iter->overlaps);
     mp_free(iter->mp, iter->dummy);
     mp_free(iter->mp, iter->head);
```

I'm still wondering why though - what am I missing?  It's such an obvious thing to do, especially given there is substantially more work to do to double check if we're leaking memory and to report it, than simply call the free function instead.  So I must be missing something - like a case where it is intended to destroy the iterator while keeping the pileup array intact still - so I'm wary!  This change does past tests, but it's very odd to be done this way.

I also see a lot of code carefully avoiding triggering it. Eg in samtools/bam2depth.c we have:

```
    while ((ret=bam_mplp_auto(mplp, &tid, &pos, n_plp, plp)) > 0) { // come to the next covered position
        if (pos < beg || pos >= end) continue; // out of range; skip
```

It is naturally more efficient to bail out on the pos >= 0 end condition so we don't have lots of needless looping to do, eg:

```
if (pos < beg) continue;
if (pos >= end) break;
```

However that triggers the memory leak, so instead it is content to waste cpu cycles instead by a non-obvious strategy of continuing even though we know we can no longer emit any output.  Again, this is deliberate, so what do I miss here?  Using the break form and a bam_plp_reset call seems to work for me.

**Comment by jkbonfield** [comment]
Well, it may have always been thus, but even then it had the bam_plp_reset function just a few lines above and yet deliberately chose to check and report a user failure rather than to simply call it and free.

@lh3 - what's the reason for not freeing here?

**Comment by jmarshall** [comment]
Hmmm… `bam_plp_reset()` has also existed since forever.  I think this is just a case of joining dots that were never previously connected.

I don't think you're missing anything; `bam_plp_destroy()` should act as if it has called `bam_plp_reset()` itself.  In fact the only thing that `reset` does that `destroy` doesn't already do is `mp_free` the whole linked list.

So we should:

1. Verify that `reset` and `destroy` appropriately clear (resp. destroy) the overlaps, and lift `reset`'s `overlap_remove()` out of the loop;
2. Change `destroy` to `mp_free` the whole linked list;
3. Verify that both `reset` and `destroy` are doing the right thing with `dummy`.

**Comment by jmarshall** [comment]
As for

        while ((ret=bam_mplp_auto(mplp, &tid, &pos, n_plp, plp)) > 0) { // come to the next covered position
        if (pos < beg || pos >= end) continue; // out of range; skip

continuing rather than breaking seems to me like sensible defensive programming, unrelated to avoiding triggering the memory leak.  If the iterator stuff isn't quite working right, client code just skips an out of range position and still processes any following in-range ones.  (Hmmm… `pos` comes straight from `bam_mplp_auto` which surely returns them sequentially.  So it's somewhat excessive defensiveness :smile:, but I still think it's unrelated to `reset` vs. `destroy`.)

**Comment by jkbonfield** [comment]
In ```bam_plp_init()``` head and tail both point to the same ```mp_alloc()``` result and dummy to its own ```mp_alloc()``` result, so it starts with 2 memory allocations, analogous to the 2 deallocations in ```bam_plp_destroy()```.

However ```bam_plp_reset()``` will ```mp_free(iter->head)``` so destroy would then fail I think.   So my trivial fix isn't trivial afterall.  Ugh this is a can of worms!

**Comment by jkbonfield** [comment]
My mistake.  ```bam_plp_reset()``` loop keeps going until p->next is NULL, so it always stops with one item left (the tail).  ```bam_plp_push()``` (almost) always creates a new tail (with the exception of two sequences starting at the same point and the second sequence being 0 bases long).  Therefore I think it's fair to say that under valid data it'll always be acceptable to reset before destroying.

The dummy entry never gets reassigned.  It is simply used with dummy->next in order to hold a pointer temporarily.  Infact it doesn't need to exist at all and could just be a local pointer in ```bam_plp_next```.  Therefore I think that memory free is just fine.

In addition to this I've tried all sorts of testing using valgrind and a variety of manually made sam files and it seems to not leak or double free anywhere, so I think my analysis is OK.

So in summary, the only changes needed would be bam_plp_reset() call in bam_plp_destroy() and moving the overlap_remove() call in bam_plp_reset() outside the loop.  I don't think destroy should directly do the mp_frees as we may as well just use reset for that so it's only in one place.

The question is, do we trust this for 1.3?  Reward for fixing is neglible, vs risk for mistake being high.  IMO wait and fix at start of 1.4.

**Comment by jmarshall** [comment]
> The dummy entry never gets reassigned

Apologies, I misread [`iter->head = iter->dummy->next`](https://github.com/samtools/htslib/blob/develop/sam.c#L1861) as prepending `dummy` to the list.  So actually it's just used in `bam_plp_next()` for tracking the previous node so it can delete an item from the list — and this could be recoded to use a `lbnode_t **prev_next` instead of a whole dummy node if we wished.

> I don't think destroy should directly do the mp_frees as we may as well just use reset for that so it's only in one place.

It's doing a different job from `reset`: in particular, it needs to free the tail node too.  So IMHO there's a case for its own separate code (which it has 90% of already).

**Comment by jkbonfield** [comment]
It doesn't need to free tail.  Both head and tail on creation point to the same object.  After reset both head and tail once again point to the same object.  Therefore it is correct to only free one of them.  I have verified this with Valgrind too.

I don't particularly mind calling reset as part of destroy as it's not so hard to duplicate the 7 or so lines, but that's most of the function so I don't really see what the duplication gains us.  Anyway, if we're not going to do this for 1.3 (and I'd be hestitant to keep adding small things to the merge-list) then we can discuss it more later.

**Comment by jmarshall** [comment]
`reset` frees _N-1_ nodes, excluding the tail.  `destroy` frees _N_ nodes.  So if `destroy` calls `reset`, it still has one node to free itself (as we see in your PR).  The real question is: which leads to easier to comprehend code in `bam_plp_destroy()`?  IMHO "free all items in linked list" is clearer than "call `reset` (what else does it do?), then free the head item again (?!)", but YMMV…

**Comment by jmarshall** [comment]
Spotted in the wild: https://www.biostars.org/p/173354/
